pkg/server/handler/sso_customtoken.go

```diff
@@ -0,0 +1,294 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"github.com/dgrijalva/jwt-go"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/skygeario/skygear-server/pkg/server/asset"
+	"github.com/skygeario/skygear-server/pkg/server/authtoken"
+	"github.com/skygeario/skygear-server/pkg/server/plugin/hook"
+	"github.com/skygeario/skygear-server/pkg/server/plugin/provider"
+	"github.com/skygeario/skygear-server/pkg/server/router"
+	"github.com/skygeario/skygear-server/pkg/server/skydb"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+)
+
+type ssoCustomTokenClaims struct {
+	Profile skydb.Data `json:"skyprofile"`
+	jwt.StandardClaims
+}
+
+type ssoCustomTokenLoginPayload struct {
+	keyFunc     jwt.Keyfunc
+	TokenString string               `mapstructure:"token"`
+	Token       *jwt.Token           `mapstructure:"-"`
+	Claims      ssoCustomTokenClaims `mapstructure:"-"`
+}
+
+func (payload *ssoCustomTokenLoginPayload) Decode(data map[string]interface{}) skyerr.Error {
+	if err := mapstructure.Decode(data, payload); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	}
+
+	if parsedToken, err := jwt.ParseWithClaims(
+		payload.TokenString,
+		&payload.Claims,
+		payload.keyFunc,
+	); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	} else {
+		payload.Token = parsedToken
+	}
+
+	return payload.Validate()
+}
+
+func (payload *ssoCustomTokenLoginPayload) Validate() skyerr.Error {
+	claims := payload.Claims
+	if claims.Subject == "" {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: subject (sub) not specified",
+		)
+	}
+
+	if claims.ExpiresAt == 0 {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: expires at (exp) not specified",
+		)
+	}
+
+	if claims.IssuedAt == 0 {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: issued at (iat) not specified",
+		)
+	}
+
+	if claims.Valid() != nil {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: token is not valid at this time",
+		)
+	}
+
+	return nil
+}
+
+/*
+SSOCustomTokenLoginHandler authenticate user with password
+
+The user can be either identified by username or password.
+
+curl -X POST -H "Content-Type: application/json" \
+  -d @- http://localhost:3000/ <<EOF
+{
+	"action": "sso:custom_token:login",
+	"token": "eyXXXXXXX",
```

Comment:
I think the purpose of the example is to provide a minimal working example with the default setting.

I suggest the following:

1. Add the setting with default to `.env.sample`
2. the token here is actually valid with default setting.

Reply:
even if the token is correctly signed the request still wouldn’t be valid because the validity period will cause the token to become invalid

but I totally recognise we need a better example, so i will also include info on how this token is generated